tool/tbot/tshwrap/wrap.go

```diff
@@ -0,0 +1,327 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tshwrap
+
+import (
+	"encoding/json"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/teleport/api/identityfile"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/tool/tbot/config"
+	"github.com/gravitational/teleport/tool/tbot/identity"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// TSHVarName is the name of the environment variable that can override the
+	// tsh path that would otherwise be located on the $PATH.
+	TSHVarName = "TSH"
+
+	// TSHMinVersion is the minimum version of tsh that supports Machine ID
+	// proxies.
+	TSHMinVersion = "9.3.0"
+)
+
+var log = logrus.WithFields(logrus.Fields{
+	trace.Component: teleport.ComponentTBot,
+})
+
+// TSHRunner is a method for executing tsh.
+type TSHRunner interface {
+	// Capture runs tsh and captures its standard output.
+	Capture(args ...string) ([]byte, error)
+
+	// exec runs tsh with the given additional environment variables and args.
+	// It should inherit the current process's input and output streams.
+	Exec(env map[string]string, args ...string) error
+}
+
+type execTSHRunner struct {
+	path string
+}
+
+func (r *execTSHRunner) Capture(args ...string) ([]byte, error) {
+	out, err := exec.Command(r.path, args...).Output()
+	if err != nil {
+		return nil, trace.Wrap(err, "error executing tsh")
+	}
+
+	return out, nil
+}
+
+func (r *execTSHRunner) Exec(env map[string]string, args ...string) error {
+	// The subprocess should inherit the environment plus our vars.
+	environ := os.Environ()
+	for k, v := range env {
+		environ = append(environ, k+"="+v)
+	}
+
+	log.Debugf("executing %s with env=%+v and args=%+v", r.path, env, args)
+
+	child := exec.Command(r.path, args...)
+	child.Env = environ
+	child.Stdin = os.Stdin
+	child.Stdout = os.Stdout
+	child.Stderr = os.Stderr
+
+	return trace.Wrap(child.Run(), "unable to execute tsh")
+}
+
+// NewRunner attempts to locate a `tsh` binary on the local system. The
+// standard path lookup behavior can be overridden by setting the `$TSH`
+// environment variable to point to a different tsh executable.
+func NewRunner() (TSHRunner, error) {
+	if val, ok := os.LookupEnv(TSHVarName); ok {
+		return &execTSHRunner{
+			path: val,
+		}, nil
+	}
+
+	binary := "tsh"
+	if runtime.GOOS == constants.WindowsOS {
+		binary = "tsh.exe"
+	}
+
+	path, err := exec.LookPath(binary)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &execTSHRunner{
+		path: path,
+	}, nil
+}
+
+// GetTSHVersion queries the system tsh for its version.
+func GetTSHVersion(r TSHRunner) (*semver.Version, error) {
```

Comment:
I think it makes some sense to keep this uncoupled. The only other wrapper command I (maybe) anticipate is `tsh ssh`, which has no version dependency at all, meaning it wouldn't need to call `tsh version` and risk erroring out on a pre-`-f json` tsh that would otherwise work fine.

Reply:
Fair enough.